File: lab5/exp.py
num = []
op = []

exp = "((-1)+(-4-(2+3)))"
flag = False

def cal(operator,num1,num2):
    if(operator == '+'):
        return int(num1) + int(num2)
    else:
        return int(num2)-int(num1)
before = ' '
for i in exp:
    if(i.isdigit()):
        if (flag):
            num.append(-int(i))
            flag = False
        else:
            num.append(int(i))
    elif(i=='(' or i == '+'):
        op.append(i)
    elif(i==')'):
        while(op[-1]!='('):
            operator = op.pop(-1)
            num1 = num.pop(-1)
            num2 = num.pop(-1)
            num.append(cal(operator,num1,num2))
        op.pop(-1)
        
    else:
        if(before =='('):
            flag = True
        else:
            op.append('-')
    before = i
        

File: lab5/exp.c
#include<stdio.h>
int calop(char op,int num1,int num2){
	if(op=='+')
		return num1+num2;
	else
		return num2-num1;

}
int isdig(char i){
	if (i<='9'&&i>='0')
		return 1;
	else
		return 0;	
}
int main(){
	char exp[20];
	int num[20];
	char op[20];
	int ptrnum=0;
	int ptrop=0;
	int flag = 0;
	char before = '0';
	scanf("%s",exp);
	for (int i=0;exp[i];i++){
		if(isdig(i))
			if(flag){
				flag = 0;
				num[ptrnum++] = -(i-'0');
			}
			else
				num[ptrnum++]=i-'0';
		else if(i=='(' || i=='+')
			op[ptrop++]=i;
		else if(i==')'){
			if(op[ptrop]!='('){	//这里不写while是因为最多仅一括号
				char cal = op[ptrop--];
				int num1 = num[ptrnum--];
				int num2 = num[ptrnum--];
				num[ptrnum++] = calop(cal,num1,num2);
			}
			op[ptrop--]='\0';
		}
		else{
			if(before == '(')
				flag = 1;
			else
				op[ptrop--]='-';
		
		}
		before = i;	
		
	}

	printf("%d\n",num[0]);
}
